### DISCUSSION TITLE
Zig 0.11 includes a basic http server

### DISCUSSION THREAD

**Discussion by yerlaser** [zigzap/zap#34] [General]
Hi,
Zig 0.11 has been released and according to merged commits it seems to include a basic http server.
Have you seen it?

**Comment by renerocksai** [comment]
Of course I have. 

Since I don't know the purpose of your yes / no question, I'm trying to give you my opinion:

I wouldn't use the std http server in production for my own purposes, though. I think was mainly made for testing the http client. It's far from being proven and battle-tested, and as zig currently still lacks async, the most straight-forward way to use it in a multi-connection ( = realistic) scenario is to use one thread per connection which is not ideal, especially when dealing with thousands of concurrent connections.

Facil.io comes with a full-blown async IO reactor framework that supports multiple worker processes and worker threads, among many other things.

So, serving HTTP is one thing, but the underlying IO code is another.

Now, once Zig reintroduces async, I would hope for (and expect) cool http server frameworks to pop up / be revived. That would be really interesting. Especially, since their most compelling argument to me would be that they'd be written entirely in Zig.

But I need something now, that's not a gamble ("battle tested"), is feature-rich, and can handle 10k parallel connections with ease.

The author of facil.io is working on a new version that's even more compelling since it will support things like http streaming without blocking a worker thread. I guess, I will incorporate that into zap then - but will also keep an eye on upcoming zig implementations.

**Reply by yerlaser** [reply]
Thanks for the very detailed answer.
Yes, I mostly meant you avoid doing duplicate work and somehow reuse the parts that are already in the std lib.
But, your reply makes it clear that it's not feasible.
It's sad that so many people have to do duplicate work.

**Reply by yerlaser** [reply]
I was hoping to see Zig be suitable for embedded programming, e.g. on ESP32 chips.
I need a basic http server on it.
But, it seems I need to look elsewhere.

**Reply by renerocksai** [reply]
I have yet to try Zig on the ESP32, but the ESP32-C3 sports a RISC-V CPU which is probably already supported by MicroZig. Having only used ESP32 with the Arduino libs, I don't know how easy it is to get a custom HTTP server running on top of the WiFi. I bet Felix and Matt from the [Zig Embedded Group](https://github.com/ZigEmbeddedGroup) can shed more light on your issue. Might be interesting.

Having said what I said about std.http.server, it might be suitable as a basic HTTP server that doesn't need to handle any / too many parallel connections.

Zig is definitely well-suited for embedded programming. I attended a [Zig Embedded Workshop](https://github.com/ZigEmbeddedGroup/synth-workshop) from @mattnite recently, where we used a Raspberry Pico (RP2040). Check out their Discord, too, if you like. It's all on the [Zig Embedded site](https://microzig.tech/).

**Reply by yerlaser** [reply]
Yeah, I am monitoring ZEG. It's a great initiative. Unfortunately, there are only two people working on it atm in their free time and the progress is understandably slow.

**Reply by renerocksai** [reply]
> Yeah, I am monitoring ZEG. It's a great initiative. Unfortunately, there are only two people working on it atm in their free time and the progress is understandably slow.

Make it three then 😉!

But I get your point.
